alphagov/paas-bootstrap
# [#130102001] Bootstrap bosh and concourse

# What

Story: [Create custom release concourse](https://www.pivotaltracker.com/story/show/130102001)

We need a new concourse that will be used to build bosh releases. It was decided to take the opportunity and deploy this concourse with bosh instead of bosh-init. The main reasons are:

* bosh-init can handle only 1 VM. bosh can handle complex deployments and we may need it to scale concourse
* bosh can cache compiled artifacts
* bosh manages the state automatically
* reduces the size of the CF pipeline so this should improve our development time

This is now a generic concourse that we will use to build bosh releases, but eventually paas-cf may be migrated to use it for its bootstrap. We created a new repository that uses a lot of the paas-cf code so most is already known. The main changes are:

* bosh is now created in the bootstrap pipeline instead of the CF pipeline
* concourse is deployed with bosh instead of bosh-init
* since concourse lite is in a different VPC than bosh, bosh is now accessed via its external IP
* the security groups have changed slightly because we moved things around and the dependencies have changed
* update CPI to latest (v60). This was required for the auto-assign feature introduced in v55. The pipeline point to the bosh-init container tagged for the new version.
* changed IAM policies (See [PR in aws-account-wide-terraform](https://github.gds/government-paas/aws-account-wide-terraform/pull/58))

We have made other smaller changes:

* improve naming
* parallelise and merge pipeline jobs
* improve management of bosh_fqdn variable
* other minor things

# How to review
* Apply changes from the [PR in aws-account-wide-terraform](https://github.gds/government-paas/aws-account-wide-terraform/pull/58)
* Start concourse lite:

```
DEPLOY_ENV=testabcd make dev bootstrap
```

* Connect to `localhost:8080` and start `create` pipeline
* In the end you should be able to access concourse at: `https://concourse.testabcd.dev.cloudpipeline.digital`
* Login to concourse, upload pipelines
* Destroy the environment using the destroy pipeline
* Destroy concourse-lite:

```
DEPLOY_ENV=testabcd make dev bootstrap-destroy
```

* Set up the local environment with python, ruby, go and check that `make test` passes successfully

# Create the build CI

* Make sure the previous concourse lite is destroyed
* Start concourse lite for CI:
```
DEPLOY_ENV=build make ci bootstrap
```
* Connect to `localhost:8080` and start `create` pipeline
* In the end you should be able to access concourse at: `https://concourse.build.ci.cloudpipeline.digital`
* Destroy concourse-lite:

```
DEPLOY_ENV=build make ci bootstrap-destroy
```

# Who can review
Anyone but @HenryTK or myself
